cmd/internal/cli/inspect_linux.go

```diff
@@ -128,111 +170,109 @@ var InspectCmd = &cobra.Command{
-		delimiter := "@@@end"
 
 		if helpfile {
 			sylog.Debugf("Inspection of helpfile selected.")
-
-			// append to a[2] to run commands in container
-			a[2] += fmt.Sprintf(" echo '%v\nhelpfile';", prefix)
-			a[2] += getHelpFile(AppName)
-			a[2] += fmt.Sprintf(" echo '%v';", delimiter)
+			a[2] += getHelpCommand(AppName)
 		}
 
 		if deffile {
 			sylog.Debugf("Inspection of deffile selected.")
-
-			// append to a[2] to run commands in container
-			a[2] += fmt.Sprintf(" echo '%v\ndeffile';", prefix)
-			a[2] += " cat .singularity.d/Singularity;" // apps share common definition file
-			a[2] += fmt.Sprintf(" echo '%v';", delimiter)
+			a[2] += getDefinitionCommand()
 		}
 
 		if runscript {
 			sylog.Debugf("Inspection of runscript selected.")
-
-			// append to a[2] to run commands in container
-			a[2] += fmt.Sprintf(" echo '%v\nrunscript';", prefix)
-			a[2] += getRunscriptFile(AppName)
-			a[2] += fmt.Sprintf(" echo '%v';", delimiter)
+			a[2] += getRunscriptCommand(AppName)
 		}
 
 		if testfile {
 			sylog.Debugf("Inspection of test selected.")
-
-			// append to a[2] to run commands in container
-			a[2] += fmt.Sprintf(" echo '%v\ntest';", prefix)
-			a[2] += getTestFile(AppName)
-			a[2] += fmt.Sprintf(" echo '%v';", delimiter)
+			a[2] += getTestCommand(AppName)
 		}
 
 		if environment {
 			sylog.Debugf("Inspection of environment selected.")
-
-			// append to a[2] to run commands in container
-			a[2] += fmt.Sprintf(" echo '%v\nenvironment';", prefix)
-			a[2] += getEnvFile(AppName)
-			a[2] += fmt.Sprintf(" echo '%v';", delimiter)
+			a[2] += getEnvironmentCommand(AppName)
 		}
 
-		// default to labels if nothing was appended
+		// Default to labels if nothing was appended
 		if labels || len(a[2]) == 0 {
-			sylog.Debugf("Inspection of labels as default.")
-
-			// append to a[2] to run commands in container
-			a[2] += fmt.Sprintf(" echo '%v\nlabels';", prefix)
-			a[2] += getLabelsFile(AppName)
-			a[2] += fmt.Sprintf(" echo '%v';", delimiter)
+			sylog.Debugf("Inspection of labels selected.")
+			a[2] += getLabelsCommand(AppName)
 		}
 
+		// Execute the compound command string.
 		fileContents, err := getFileContent(abspath, name, a)
 		if err != nil {
-			sylog.Fatalf("While getting helpfile: %v", err)
+			sylog.Fatalf("Could not inspect container: %v", err)
 		}
 
-		contentSlice := strings.Split(fileContents, delimiter)
-		for _, s := range contentSlice {
-			s = strings.TrimSpace(s)
-			if strings.HasPrefix(s, prefix) {
-				split := strings.SplitN(s, "\n", 3)
-				if len(split) == 3 {
-					attributes[split[1]] = split[2]
-				} else if len(split) == 2 {
-					sylog.Warningf("%v metadata was not found.", split[1])
+		inspectObj := inspectFormat{}
+		inspectObj.Type = "container"
+		inspectObj.Attributes.Labels = make(map[string]string)
+		inspectObj.Attributes.Environment = make(map[string]string)
+
+		// Parse the command output string into sections.
+		reader := bufio.NewReader(strings.NewReader(fileContents))
```

Comment:
you might want a bufio.Scanner here (see https://golang.org/pkg/bufio/#example_Scanner_lines)

Reply:
Thanks! Fixed.

Reply:
I'm not sure Scanner will work based on my parsing structure. I alternate between reading one line and reading X bytes where the X is determined based on the line that was read (the actual number of bytes for the section, as returned by `wc -c`). Scanner doesn't have a straight-forward construct to read X bytes as far as I can tell.

Reply:
What do you think @mem ?

Reply:
Yeah, writing a split function to handle this case is probably not worth it because you would have to add state.